File: FastMines_android/src/main/java/fmg/android/app/model/items/MosaicDataItem.java
package fmg.android.app.model.items;

import androidx.databinding.Bindable;

import java.util.Objects;

import fmg.android.app.BR;
import fmg.android.img.AndroidBitmapView;
import fmg.android.img.MosaicImg2;
import fmg.common.geom.BoundDouble;
import fmg.common.geom.Matrisize;
import fmg.core.img.MosaicImageModel2;
import fmg.core.types.EMosaic;
import fmg.core.types.ESkillLevel;

import static fmg.core.img.PropertyConst.PROPERTY_MOSAIC_TYPE;
import static fmg.core.img.PropertyConst.PROPERTY_SIZE;
import static fmg.core.img.PropertyConst.PROPERTY_SKILL_LEVEL;

/** Mosaic item for data model */
public class MosaicDataItem extends BaseDataItem<EMosaic, MosaicImageModel2, AndroidBitmapView<MosaicImageModel2>, MosaicImg2.MosaicAndroidBitmapController> {

    private ESkillLevel skillLevel;

    public MosaicDataItem(EMosaic mosaicType) {
        super(mosaicType);
        setTitle(fixTitle(mosaicType));
    }

    @Bindable
    public EMosaic getMosaicType()                   { return getUniqueId(); }
    public void    setMosaicType(EMosaic mosaicType) {        setUniqueId(mosaicType); }

    @Bindable
    public ESkillLevel getSkillLevel() {
        if (skillLevel == null)
            setSkillLevel(ESkillLevel.eBeginner);
        return skillLevel;
    }
    public void setSkillLevel(ESkillLevel skillLevel) {
        Objects.requireNonNull(skillLevel, "Value of type " + ESkillLevel.class.getSimpleName() + " must be defined");
        if (this.skillLevel == skillLevel)
            return;
        this.skillLevel = skillLevel;
        firePropertyChanged(PROPERTY_SKILL_LEVEL);
    }

    @Override
    public MosaicImg2.MosaicAndroidBitmapController getEntity() {
        if (this.entity == null) {
            Matrisize sizeField = getSkillLevel().sizeTileField(getMosaicType());
            var tmp = new MosaicImg2.MosaicAndroidBitmapController();
            var m = tmp.getModel();
            m.setMosaicType(getMosaicType());
            m.setSizeField(sizeField);
            m.setPadding(new BoundDouble(5 * getZoom()));
            m.setRotateMode(MosaicImageModel2.ERotateMode.SOME_CELLS);
//            m.setBackgroundColor(MosaicDrawModel.DefaultBkColor);
            m.getPenBorder().setWidth(3 * getZoom());
//            m.setRotateAngle(45 * ThreadLocalRandom.current().nextInt(7));

//            android.graphics.BitmapView bmp = tmp.getImage();
//            assert (bmp.getWidth()  == (int)(getSize().width * getZoom()));
//            assert (bmp.getHeight() == (int)(getSize().height * getZoom()));
            setEntity(tmp);
        }
        return this.entity;
    }

    @Override
    protected void firePropertyChanged(String propertyName) {
        if (isDisposed())
            return;

        super.firePropertyChanged(propertyName);

        switch (propertyName) {
        case PROPERTY_SIZE: // TODO delete this case!!
            super.firePropertyChanged(PROPERTY_TITLE);
            break;
        case PROPERTY_UNIQUE_ID:
            getEntity().getModel().setMosaicType(getMosaicType());
            getEntity().getModel().setSizeField(calcSizeField(getSkillLevel()));
            setTitle(fixTitle(getMosaicType()));
            break;
        case PROPERTY_SKILL_LEVEL:
            getEntity().getModel().setSizeField(calcSizeField(getSkillLevel()));
            break;
        case PROPERTY_MOSAIC_TYPE:
            super.firePropertyChanged(PROPERTY_UNIQUE_ID);
            break;
        }
    }

    @Override
    protected void firePropertyChangedAsync(String propertyName) {
        if (isDisposed())
            return;

        super.firePropertyChangedAsync(propertyName);

        // refire as android data binding event
        switch (propertyName) {
        case PROPERTY_MOSAIC_TYPE: notifyPropertyChanged(BR.mosaicType); break;
        case PROPERTY_SKILL_LEVEL: notifyPropertyChanged(BR.skillLevel); break;
        }
    }

    private static String fixTitle(EMosaic mosaicType) {
        return mosaicType.getDescription(false);//.replace("-", "\u2006-\u2006");
    }

    private Matrisize calcSizeField(ESkillLevel skill) {
        return ((skill == ESkillLevel.eCustom)
                    ? ESkillLevel.eBeginner
                    : skill)
                .sizeTileField(getMosaicType());
    }

}

File: FastMines_uwp/FastMines.Uwp.App/Model/Items/MosaicDataItem.cs
using System.ComponentModel;
using Microsoft.Graphics.Canvas;
using Fmg.Common.Geom;
using Fmg.Core.Types;
using Fmg.Core.Img;
using MosaicModel      = Fmg.Core.Img.MosaicAnimatedModel<Fmg.Common.Nothing>;
using MosaicView       = Fmg.Uwp.Img.Win2d.MosaicImg.CanvasBmpView;
using MosaicController = Fmg.Uwp.Img.Win2d.MosaicImg.CanvasBmpController;

namespace Fmg.Uwp.App.Model.Items {

    /// <summary> Mosaic item for data model </summary>
    public class MosaicDataItem : BaseDataItem<EMosaic, MosaicModel, MosaicView, MosaicController> {

        private ESkillLevel _skillLevel;

        public MosaicDataItem(EMosaic mosaicType)
            : base(mosaicType)
        {
            Title = FixTitle(mosaicType);
        }

        public EMosaic MosaicType {
            get => UniqueId;
            set {  UniqueId = value; }
        }

        public ESkillLevel SkillLevel {
            get { return _skillLevel; }
            set {
                _notifier.SetProperty(ref _skillLevel, value);
            }
        }

        public override MosaicController Entity {
            get {
                if (entity == null) {
                    var sizeField = SkillLevel.SizeTileField(MosaicType);
                    var tmp = new MosaicController(CanvasDevice.GetSharedDevice());
                    var m = tmp.Model;
                    m.MosaicType = MosaicType;
                    m.SizeField = sizeField;
                    m.Padding = new BoundDouble(5 * Zoom());
                    m.RotateMode = EMosaicRotateMode.someCells;
                    //m.BackgroundColor = MosaicDrawModelConst.DefaultBkColor;
                    m.PenBorder.Width = 3 * Zoom();
                    //m.RotateAngle = 45 * ThreadLocalRandom.Current.Next(7);

                    //var bmp = tmp.Image;
                    //System.Diagnostics.Debug.Assert(bmp.SizeInPixels.Width  == (int)(Size.Width  * Zoom()));
                    //System.Diagnostics.Debug.Assert(bmp.SizeInPixels.Height == (int)(Size.Height * Zoom()));
                    Entity = tmp; // call this setter
                }
                return entity;
            }
        }

        protected override void OnPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            base.OnPropertyChanged(sender, ev);
            switch(ev.PropertyName) {
            case nameof(this.UniqueId):
                _notifier.FirePropertyChanged(nameof(this.MosaicType)); // recall with another property name
                Entity.MosaicType = MosaicType;
                Entity.SizeField = CalcSizeField(SkillLevel);
                Title = FixTitle(MosaicType);
                break;
            case nameof(this.SkillLevel):
                Entity.SizeField = CalcSizeField(SkillLevel);
                break;
            }
        }

        private static string FixTitle(EMosaic mosaicType) {
            return mosaicType.GetDescription(false);//.Replace("-", "\u2006-\u2006");
        }

        private Matrisize CalcSizeField(ESkillLevel skill) {
            return ((skill == ESkillLevel.eCustom)
                        ? ESkillLevel.eBeginner
                        : skill)
                    .SizeTileField(MosaicType);
        }

    }

}
